Modix.Bot/Behaviors/PromotionPollBehavior/PromotionDialogBehavior.cs

```diff
@@ -0,0 +1,211 @@
+
+            await message.AddReactionsAsync(new[] {(IEmote)_approveEmoji, _disapproveEmoji});
+
+            var poll = new CachedPromoDialog
+            {
+                MessageId = message.Id,
+                CampaignId = campaignSummary.Id,
+            };
+
+            SetDialogCache(campaign.Campaign.Id, message.Id, poll);
+            await _promotionDialogRepository.CreateAsync(message.Id, campaign.Campaign.Id);
+        }
+
+        private async Task DeletePromoDialog(PromotionActionCreationData campaign)
+        {
+            if (!_memoryCache.TryGetValue(campaign.Campaign.Id, out CachedPromoDialog poll))
+                return;
+            var pollChannel = await GetPollChannel(_discordSocketClient.GetGuild(campaign.GuildId));
+
+            var message =  await pollChannel.GetMessageAsync(poll.MessageId);
+
+            if (!await _promotionDialogRepository.TryDeleteAsync(message.Id))
+                throw new InvalidOperationException("Dialog to be deleted does not exist");
+
+            await message.DeleteAsync();
+        }
+
+        private EmbedBuilder BuildPollEmbed(PromotionCampaignSummary campaign, int Approve, int Oppose, string error = null)
+        {
+                var boldRole = $"**{MentionUtils.MentionRole(campaign.TargetRole.Id)}**";
+                var user = _discordSocketClient.GetUser(campaign.Subject.Id);
+
+                ValueTask<Color> aviColor;
+
+                if ((user.GetAvatarUrl(size: 16) ?? user.GetDefaultAvatarUrl()) is { } avatarUrl)
+                {
+                    aviColor = _imageService.GetDominantColorAsync(new Uri(avatarUrl));
+                }
+                else
+                {
+                    aviColor = new ValueTask<Color>(Color.Teal);
+                }
+
+                var embed = new EmbedBuilder()
+                    .WithTitle($"#{campaign.Id}")
+                    .WithDescription($"**{campaign.Subject.Nickname}** Has been nominated for promotion to {boldRole} ")
+                    .WithColor(aviColor.Result)
+                    .WithUserAsAuthor(user)
+                    .AddField("The current vote total stands at: ", $" {Approve} {_approveEmoji} / {Oppose} {_disapproveEmoji}")
+                    .WithTimestamp(_utcNow)
+                    .WithFooter("Please react with the appropriate reaction to cast your vote!");
+
+                return error == null ? embed : embed.AddField("⚠️ Vote Error: ⚠️", error);
+        }
+
+        private async Task<ITextChannel> GetPollChannel(IGuild guild)
+        {
+            var getPollChannel = await _designatedChannelService
+                .GetDesignatedChannelsAsync(guild, DesignatedChannelType.PromotionInterface);
+            return getPollChannel.First() as ITextChannel;
+        }
+
+        private void SetDialogCache(long campaign, ulong message, CachedPromoDialog dialog)
+        {
+            _memoryCache.Set(campaign, dialog);
```

Comment:
BIG RED FLAG!

Your keys for `IMemoryCache` have to be GLOBALLY unique.

Reply:
are campaign ids not unique?

Reply:
Campaign IDs are unique *within the context of campaigns*. But the memory cache is a global singleton throughout the app, so just using the campaign ID can cause collisions with anything else that uses the memory cache.

You can find a few examples elsewhere in MODiX, but we've usually been following a pattern of doing something like the following to get unique keys:
```cs
private object GetKey(ulong id) => new { Target = "NameOfBehavior", id };
```

Reply:
That is actually MASSIVELY WRONG. Anonymous objects do not implement value-equality semantics. That will mean we fill up the cache with bogus entries that NEVER get used.

Your keys have to be globally unique AMONG EVERY CLASS IN THE ENTIRE APPLICATION THAT USES THE CACHE.

Reply:
this should be a ValueTuple

Reply:
No, I'm pretty sure Jake's MASSIVELY WRONG. Anonymous objects implement `Equals` and `GetHashCode` based on the values of their properties.

Reply:
I stand corrected. Since `IMemoryCache` compares keys via `.Equals()` rather than `==`, it works fine. Anonymous object types do implement value equality semantics for `.Equals()`.